presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,43 +582,193 @@ private static NestedField nestedField(String name)
+                        // If function transforms the shape of the input map or array argument and the form of such transformation was not supplied in
+                        // getOutputToInputTransformationFunction() then we do not know how to attribute those subfields to the input argument. Therefore, we have to give up on
+                        // pruning unaccessed subfields and read all of them.
+                        context.giveUpOnCollectingLambdaSubfields();
+                    }
+                    if (!functionDescriptor.isReturningInputValues()) {
+                        context.startCollectingLambdaSubfields();
+                    }
 
-                // visit subscript expressions only
-                Optional<Subfield> subfield = toSubfield(call, functionResolution, expressionOptimizer, connectorSession);
-                if (subfield.isPresent()) {
-                    context.subfields.add(subfield.get());
+                    transformOuterLambdaSubfieldsIfNeeded(context, functionDescriptor);
+
+                    if (context.getLambdaSubfields().isPresent()) {
+                        Set<Integer> argumentIndicesContainingMapOrArray = functionDescriptor.getArgumentIndicesContainingMapOrArray()
+                                .orElse(IntStream.range(0, call.getArguments().size()).boxed().collect(Collectors.toSet()));
+
+                        // All the lambda subfields collected in outer functions relate only to the arguments of the function specified in
+                        // functionDescriptor.argumentIndicesContainingMapOrArray.
+                        // argumentIndexToLambdaSubfieldsMap contains the mapping of the function argument and lambda subfields related to it. Initially, it contains only lambda
+                        // subfields from outer functions.
+                        Map<Integer, Set<Subfield>> argumentIndexToLambdaSubfieldsMap = argumentIndicesContainingMapOrArray.stream()
+                                .collect(Collectors.toMap(callArgumentIndex -> callArgumentIndex, unused -> new HashSet<>(context.getLambdaSubfields().get())));
+
+                        // If the function accepts lambdas, add all the lambda subfields from each lambda.
+                        for (LambdaDescriptor lambdaDescriptor : functionDescriptor.getLambdaDescriptors()) {
+                            if (!addLambdaSubfields(call, context, lambdaDescriptor, argumentIndexToLambdaSubfieldsMap)) {
+                                context.giveUpOnCollectingLambdaSubfields();
+                                call.getArguments().forEach(argument -> argument.accept(this, context));
+                                return null;
+                            }
+                        }
+
+                        if (!functionDescriptor.isReturningInputValues()) {
```

Comment:
We shouldn't need this if condition around `addNoneSubfieldIfNoAccessedSubfieldsFound`.  if it is returning input values (and doesn't have a transformation function per the knew proposed definition), we should have handled this case already by indicating all subfields were accessed.

Reply:
removed `isReturningInputValues` in separate commit that addresses this comment.